### DISCUSSION TITLE
Forgotten PR triage

### DISCUSSION THREAD

**Discussion by Kobzol** [intellij-rust/intellij-rust#7511] [IntelliJ Rust Plugin Development]
Oh, I see that you have enabled GH discussions! Well, then I gotta go try them :smile:

I wanted to ask about the amount of opened PRs in the project in relation to https://github.com/intellij-rust/intellij-rust/issues/1698. The number of opened PRs has fluctuated between 90-100 in the recent months, which IMO is quite a lot. It seems to me that having that many opened PRs makes it easier for them to "rot". An old PR will necessarily eventually get stale and full of merge conflicts as the codebase evolves away. Having so many open PRs makes it easier to overlook/forget about some of them, simply because they get lost in the swamp.

It can also de-motivate contributors that do not get feedback for a long time, because their PR is simply forgotten, even though a lot of other PRs get merged in the meantime. This might be of course caused by priority if the "forgotten" PR is not high-prio, but at least this could be communicated to the PR author somehow so that he/she knowns that he/she just has to wait and that the PR was not forgotten. Even just saying that there is currently no time for reviewing a specific PR can go a long way instead of total radio silence.

Before I continue, I want to mention two things:
1) Yes, I do realize that I myself am a large contributor to the "problem" of many PRs because I spam with them a lot and a third of the currently opened PRs come from me. :laughing: And this post is not mainly about my PRs, I am quite active here, you know me already and my PRs usually get attention. But some other contributors might not get that kind of attention.
2) I know that you, the maintainers, are swamped and I know that it's not possible to miraculously review/merge/close all of these PRs overnight. You are doing great work and I really like that you include outside contributors into the project even though the plugin is backed by a company. So please don't take this as a rant, but rather as a (maybe naive) suggestion.

I know that bug and incompatibility fixes get priority, but I wanted to suggest a (one-time) pass over the opened PRs (starting from the oldest, maybe also preferring these of outside contributors, as they have a larger chance to be "forgotten"), to try and reduce their overall count, which could then result in better visibility of new PRs and potentially a smaller chance for a PR to become "forgotten" or simply overlooked. I know that it takes time and effort and it maybe even would come at a cost of e.g. delaying the implementation of a new feature, so perhaps this idea is pretty naive :)

I suspect that most PRs fall into ~3 categories:
1) Stale PRs that have a lot of merge conflicts, are no longer relevant or were actually (re)implemented in another PR along the way. These could probably be closed without much harm to reduce PR "clutter". Keeping a PR that is probably never going to get merged open is probably useless.
2) "Long-term" PRs that are complex/controversial, that wait for some upstream Rustc/Cargo feature to be implemented, Rust feature to become stabilized or plugin feature to stop being experimental, etc. These probably have to stay open.
3) PRs that were "forgotten", i.e. no reviewer ever picked them up or notifications/messages were lost and the conversation was never followed upon. Some of them could potentially be merged with not that much effort if a reviewer picked them up.

As a sample, here are some (non-recent) PRs by outside contributors (excluding me) that have a "ball on the side of the reviewer" (by my interpretation, I might be wrong):
~https://github.com/intellij-rust/intellij-rust/pull/4235~
~https://github.com/intellij-rust/intellij-rust/pull/6704~
https://github.com/intellij-rust/intellij-rust/pull/6353
~https://github.com/intellij-rust/intellij-rust/pull/6003~
~https://github.com/intellij-rust/intellij-rust/pull/6060~
~https://github.com/intellij-rust/intellij-rust/pull/6410~
https://github.com/intellij-rust/intellij-rust/pull/6705
https://github.com/intellij-rust/intellij-rust/pull/6929
https://github.com/intellij-rust/intellij-rust/pull/7134
~https://github.com/intellij-rust/intellij-rust/pull/6904~
~https://github.com/intellij-rust/intellij-rust/pull/6813~
~https://github.com/intellij-rust/intellij-rust/pull/6838~
https://github.com/intellij-rust/intellij-rust/pull/7156
~https://github.com/intellij-rust/intellij-rust/pull/7228~

There are some more that are waiting for an answer/code change from the contributor for a long time or that probably have no hope of ever being merged. These should probably be outright closed (again, to reduce PR clutter).

Well, that's about it :smile: Maybe this is a silly idea and most of the opened PRs actually fall into 2), which would mean that we cannot really reduce their count, but anyway I wanted to mention this. If it's nonsense, just consider this post to be a test of GH discussions :)

**Comment by vlad20012** [comment]
You probably guess that we are discussing the problem of a large number of PRs quite often, but we haven't come up with anything better than a simple "let's allocate more time for reviewing PRs" and so on :laughing: 
Maybe such a "one-time pass" cleanup is not a bad idea. I guess we could try.

**Reply by Kobzol** [reply]
One option how to make forgotten PRs less likely is to introduce some label system, like:
1) When a PR gets created, it is automatically labeled with "waiting-on-reviewer"
2) When a PR gets a review (or a comment), it is automatically labeled with "waiting-on-author"
3) When a PR gets a new commit, it is automatically labeled with "waiting-on-reviewer"

Then we could easily filter/search PRs that require the attention of some reviewer more easily.

**Reply by Undin** [reply]
@Kobzol we've discussed recently that having some "state" field would be great. At this moment, I see only one way to do it - via tags. But GitHub provides support for a custom fields feature as a part of their improvements in issue-tracker direction. Currently, the [feature](https://github.com/features/issues) in beta. I've already signed our organization up.
I don't want to waste time for implementing such workflow via tags to reimplement it via custom fields in nearest future.
So my plan is to have a vacation (it's not depend on situation with the feature 😄 but anyway), check availability of custom fields feature for our organization after it and decide what to do according to our state

**Reply by vlad20012** [reply]
I recently learned that I have the ability to push (and force-push) changes to PRs from forks. It's handy for old PRs :thinking:

**Reply by Kobzol** [reply]
That's a very useful feature indeed! I see that you have started to pick up, fix and merge some old PRs. Feel free to assign me to some of them, I can help with "resurrecting" these old PRs.

**Reply by vlad20012** [reply]
@Kobzol Maybe you could take look at #3790

**Reply by Kobzol** [reply]
Not even two weeks have passed and the number of opened PRs has been reduced by about 25% :) Not bad!

**Reply by vlad20012** [reply]
To be honest, I'm sooooooo tired 😩😩😩

**Reply by Kobzol** [reply]
Then take a break and rest! :) There are already 51 PRs merged for the next milestone (and it's only in the half of its "lifetime" :D ).

**Reply by Undin** [reply]
> it's only in the half of its "lifetime" :D

Actually, release branch will be created next Tuesday morning

**Reply by Kobzol** [reply]
Oh, it branches on Tuesdays... time flies :)

**Comment by Kobzol** [comment]
I recently started contributing to rustc and I liked the concept of `Waiting-for-reviewer`/`Waiting-for-author` labels in practice. Obviously using such labels won't magically reduce the number of pending PRs, but if such labels were used, it could make filtering for PRs that actually need review easier and it would be also possible to see the number of such PRs at a glance easily.

@Undin has the "custom issue fields" feature moved forward in some way? In theory these labels could exist even without any automation, by just manualy assigning them to PRs as needed.
